File: Python/0002-m-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
Time: O(n)
Space: O(1)
"""
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0)
        cur = ListNode(0)
        dummy.next = cur
        carry = 0
        while True:
            if l1:
                cur.val += l1.val
                l1 = l1.next
            if l2:
                cur.val += l2.val
                l2 = l2.next
            
            cur.val += carry
            if cur.val >= 10:
                cur.val %= 10
                carry = 1
            else:
                carry = 0
            
            if l1 or l2 or carry:
                cur.next = ListNode(0)
                cur = cur.next
            else:
                break

        return dummy.next

File: leetcode/C++/0002-m-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(0);
        ListNode* p = l1, * q = l2, * cur = dummy; 
        int sum = 0, carry = 0;
        int a = 0, b = 0;
        while (p || q) {
            a = p ? p->val : 0;
            b = q ? q->val : 0;
            sum = carry + a + b;
            carry = sum / 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            if (p) p = p->next;
            if (q) q = q->next;
        }
        if (carry > 0)
            cur->next = new ListNode(carry);
        return dummy->next;
    }
};